torch/lib/THS/generic/THSTensor.c

```diff
@@ -499,15 +513,61 @@ void THTensor_(sparseMask)(THSTensor *r_, THTensor *t, THSTensor *mask) {
     for (long i = 0; i < r_->nnz; i++) {
       long idx = 0;
       for (long d = 0; d < nDimI; d++) {
-        idx += THTensor_fastGet2d(mask_indices_, d, i) * t->stride[d];
+        long j = THTensor_fastGet2d(indices, d, i);
+        if (j < TH_INDEX_BASE || j >= TH_INDEX_BASE + t->size[d]) {
+          THError("index out of range");
+        }
+        idx += j * t->stride[d];
       }
       real val = (t->storage->data + t->storageOffset)[idx];
       THTensor_fastSet1d(r_values_, i, val);
     }
   }
+}
+
+void THTensor_(sparseCopy)(THTensor *r_, THTensor *dense, THSTensor *sparse_)
+{
+  THTensor_(resizeAs)(r_, dense);
+  THSTensor *sparse = THSTensor_(newCoalesce)(sparse_);
+
+  long k;
+  THLongTensor  *indices = THSTensor_(newIndices)(sparse);
+  THTensor      *values = THSTensor_(newValues)(sparse);
+  THLongStorage *storage = THSTensor_(newSizeOf)(sparse);
+  long          *sizes = storage->data;
+  long          nDim = THTensor_(nDimension)(dense);
+  long          nDimI = THSTensor_(nDimensionI)(sparse);
+
+  if (r_ != dense) THTensor_(copy)(r_, dense);
+
+  if (nDim > nDimI) {
+    THTensor *srcBuffer = THTensor_(new)();
+    THTensor *dstBuffer = THTensor_(new)();
+    for (k = 0; k < sparse->nnz; k++) {
+      THTensor_(set)(dstBuffer, r_);
+      for (long d = 0; d < sparse->nDimensionI; d++) {
+        THTensor_(select)(dstBuffer, dstBuffer, 0, THTensor_fastGet2d(indices, d, k));
+      }
+      THTensor_(select)(srcBuffer, values, 0, k);
+      THTensor_(copy)(dstBuffer, srcBuffer);
+    }
+    THTensor_(free)(srcBuffer);
+    THTensor_(free)(dstBuffer);
+  } else {
+    #pragma omp parallel for private(k)
```

Comment:
This is may run into some serious cache issues if you don't use `schedule(static)`. Also, it's not worth using omp for very small tensors (there's some kind of a threshold defined somewhere in TH).